framework/di/Container.php

```diff
@@ -527,7 +531,17 @@ public function resolveCallableDependencies(callable $callback, $params = [])
                 } elseif (isset(Yii::$app) && Yii::$app->has($name) && ($obj = Yii::$app->get($name)) instanceof $className) {
                     $args[] = $obj;
                 } else {
-                    $args[] = $this->get($className);
+                    // If the argument is optional we catch not instantiable exceptions.
+                    try {
+                        $args[] = $this->get($className);
+                    } catch (NotInstantiableException $e) {
```

Comment:
This will trigger a better error in case the dependency is required.

If it is optional then requiring configuration in DI doesn't make sense.